sleap_nn/inference/inference.py

```diff
@@ -0,0 +1,568 @@
+        pass
+
+    @property
+    @abstractmethod
+    def data_config(self) -> DictConfig:
+        """Function to get the data parameters from the config."""
+        pass
+
+    @abstractmethod
+    def make_pipeline(self):
+        """Function to create the data pipeline."""
+        pass
+
+    def _initialize_inference_model(self):
+        pass
+
+    def _predict_generator(self) -> Iterator[Dict[str, np.ndarray]]:
+        """Create a generator that yields batches of inference results.
+
+        This method handles creating a pipeline object depending on the model type
+        for loading the data, as well as looping over the batches and running inference.
+
+        Returns:
+            A generator yielding batches predicted results as dictionaries of numpy
+            arrays.
+        """
+        # Initialize data pipeline and inference model if needed.
+        self.make_pipeline()
+        if self.inference_model is None:
+            self._initialize_inference_model()
+
+        def process_batch(ex):
+            # Run inference on current batch.
+            preds = self.inference_model(ex)
+
+            # Add model outputs to the input data example.
+            ex.update(preds)
+
+            # Convert to numpy arrays if not already.
+            for k, v in ex.items():
+                if isinstance(v, torch.Tensor):
+                    ex[k] = ex[k].cpu().numpy()
+
+            return ex
+
+        # Loop over data batches.
+        for ex in self.data_pipeline:
+            yield process_batch(ex)
+
+    def predict(
+        self,
+        make_labels: bool = True,
+        save_labels: bool = False,
+        save_path: str = None,
+    ) -> Union[List[Dict[str, np.ndarray]], sio.Labels]:
+        """Run inference on a data source.
+
+        Args:
+            make_labels: If `True` (the default), returns a `sio.Labels` instance with
+                `sio.PredictedInstance`s. If `False`, just return a list of
+                dictionaries containing the raw arrays returned by the inference model.
+            save_labels: If `True` , saves the labels object in a `.slp` file
+            save_path: Path to save the labels file if `save_labels` is True.
+
+        Returns:
+            A `sio.Labels` with `sio.PredictedInstance`s if `make_labels` is `True`,
+            otherwise a list of dictionaries containing batches of numpy arrays with the
+            raw results.
+        """
+        # Initialize inference loop generator.
+        generator = self._predict_generator()
+
+        if make_labels:
+            # Create SLEAP data structures from the predictions.
+            pred_labels = self._make_labeled_frames_from_generator(generator)
+            if save_labels:
+                sio.io.slp.write_labels(save_path, pred_labels)
+            return pred_labels
+
+        else:
+            # Just return the raw results.
+            return list(generator)
```

Comment:
Agreed. Let's just make it implicit and assume `save_labels = True` if `save_path is not None` so that we don't have to specify two separate arguments.